### PR TITLE
RZImport Caching

### PR DESCRIPTION
**Pull request by KingOfBrian**
Enable a shared cache so fetch requests can be performed only once by loading the cache prior to starting the import. The cache is implemented as a set of dictionaries hanging off of the context's `userInfo`. The only public API expansion is `-[NSManagedObjectContext rzi_cacheAllObjectsForEntity:` and `-[NSManagedObjectContext rzi_cacheObjects: forEntity:`.

This should make it very easy to optimize very large data imports.

### CODE REVIEW

File: Extensions/NSManagedObject+RZImport.m [hunk 1]
```diff
@@ -44,69 +45,81 @@
...
+            [results addObject:importedObject];
         }];
-        
-        objects = [updatedObjects allValues];
+        return results;
     }
     else {
         // Default to creating new object instances.
-        objects = [super rzi_objectsFromArray:array withMappings:mappings];
+        return [super rzi_objectsFromArray:array withMappings:mappings];
```
Comment (alexrrouse) [inline Extensions/NSManagedObject+RZImport.m:121]: #pp - I personally prefer the single return statement rather then having multiple returns.

File: Extensions/NSManagedObject+RZImport.m [hunk 2]
```diff
@@ -381,6 +395,23 @@ + (void)rzv_logUniqueObjectsWarning
...
+        static dispatch_once_t onceToken;
+        dispatch_once(&onceToken, ^{
+            s_cachedMissingKeyClasses = [NSMutableSet set];
+        });
+
+        if ( ![s_cachedMissingKeyClasses containsObject:NSStringFromClass(self)] ) {
+            [s_cachedMissingKeyClasses addObject:NSStringFromClass(self)];
+            RZVLogInfo(@"Class %@ for entity %@ provides a primary key, but the imported primary key value is nil. This will create a new object every time.", NSStringFromClass(self), [self rzv_entityName] );
+        }
+    });
```
Comment (alexrrouse) [inline Extensions/NSManagedObject+RZImport.m:412]: #nth - Anyway we can combine some of this logic with the `rzv_logUniqueObjectsWarning`?  Maybe like a log for Type and use an enum for each of the log messages.

File: Extensions/NSManagedObject+RZImportableSubclass.h [hunk 3]
```diff
@@ -58,4 +57,11 @@
  */
 + (BOOL)rzv_shouldAlwaysCreateNewObjectOnImport;
 
+/** 
+ *  When caching managed objects, the cache will be built for the keys specified. By default `rzv_externalPrimaryKey`
+ *  is returned. Subclasses can add cache lookups for other keys by overriding this method and returning the keys.
+ *  This is useful when a different key is used elsewhere in your object graph for associations.
```
Comment (alexrrouse) [inline Extensions/NSManagedObject+RZImportableSubclass.h:63]: #pp - You may need to add an @return here.  I'm not sure what Cocoadocs looks for exactly in it's checks for documentation, but I think this lib is at 100% and we should aim to keep it there.

**Comment by arrouse** [comment]
@KingOfBrian This looks good to me.  Lets just make sure that it is fully working in an app or two or write some test cases before we merge it.

**Comment by nbonatsakis** [comment]
@KingOfBrian Looks good to me too, but I agree with @arrouse, if we can get a few tests around this it would be helpful. Minimally a test that verifies the old behavior is the default (maybe our existing tests prove that), nice to have would be tests around the caching.

**Comment by KingOfBrian** [comment]
Added some tests, and updated the performance tests to use the new `measureBlock:` API.

**Pull request closed without merging**